File: crates/sdk/src/table.rs
pub trait Table {
    type Row;

    type EventContext;

    fn count(&self) -> u64;

    fn iter(&self) -> impl Iterator<Item = Self::Row> + '_;

    type InsertCallbackId;
    fn on_insert(
        &self,
        callback: impl FnMut(&Self::EventContext, &Self::Row) + Send + 'static,
    ) -> Self::InsertCallbackId;
    fn remove_on_insert(&self, callback: Self::InsertCallbackId);

    type DeleteCallbackId;
    fn on_delete(
        &self,
        callback: impl FnMut(&Self::EventContext, &Self::Row) + Send + 'static,
    ) -> Self::DeleteCallbackId;
    fn remove_on_delete(&self, callback: Self::DeleteCallbackId);
}

pub trait TableWithPrimaryKey: Table {
    type UpdateCallbackId;
    fn on_update(
        &self,
        callback: impl FnMut(&Self::EventContext, &Self::Row, &Self::Row) + Send + 'static,
    ) -> Self::UpdateCallbackId;
    fn remove_on_update(&self, callback: Self::UpdateCallbackId);
}

File: sdks/kotlin/src/commonMain/kotlin/com/clockworklabs/spacetimedb/Table.kt
package com.clockworklabs.spacetimedb

interface Table<TRow> {
    val tableName: String
    val count: Int
    fun iter(): Sequence<TRow>
    fun onInsert(callback: (EventContext<*>, TRow) -> Unit): CallbackId
    fun removeOnInsert(id: CallbackId)
    fun onDelete(callback: (EventContext<*>, TRow) -> Unit): CallbackId
    fun removeOnDelete(id: CallbackId)
}

interface TableWithPrimaryKey<TRow> : Table<TRow> {
    fun onUpdate(callback: (EventContext<*>, TRow, TRow) -> Unit): CallbackId
    fun removeOnUpdate(id: CallbackId)
}

interface EventTable<TRow> {
    val tableName: String
    fun onInsert(callback: (EventContext<*>, TRow) -> Unit): CallbackId
    fun removeOnInsert(id: CallbackId)
}
